src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs

```diff
@@ -965,6 +960,25 @@ async Task<(bool, ImmutableArray<ProjectConfiguration>)> CheckAsync(UpToDateChec
                 // Short-lived cache of timestamp by path
                 var timestampCache = new TimestampCache(_fileSystem);
 
+                // Ensure we have a context object for the current solution build.
+                //
+                // Ordinarily, this is created when the SBM calls ISolutionBuildEventListener.NotifySolutionBuildStarting,
+                // and cleared again later when the SBM calls ISolutionBuildEventListener.NotifySolutionBuildCompleted.
+                //
+                // However there are two cases where it may be null here:
+                //
+                // 1. When performing a validation run that continues after the solution build completed, or
+                // 2. When the build occurs in response to debugging (e.g. F5) in which case the SBM calls the
+                //    FUTDC *before* it invokes any solution build events.
+                //
+                // In either case, we construct an event here lazily so that we can correctly test for the
+                // existence of copy items in CheckCopyToOutputDirectoryItems.
+                if (_solutionBuildContextProvider.CurrentSolutionBuildContext is null)
+                {
+                    _solutionBuildEventListener.NotifySolutionBuildStarting();
```

Comment:
Do we need to fire NotifySolutionBuildCompleted event also at some point later?

Reply:
Good question but no I think it's fine as is. I have a revision of this PR in the works which will be a bit cleaner though.